### PR TITLE
RA-1469: Create a new Login Page for reference application

### PR DESCRIPTION
**Pull request by Charpell**
## JIRA TICKET NAME
[RA-1469: Create a new Login Page for reference application ](https://issues.openmrs.org/browse/RA-1469)

## SUMMARY:
Create a new Login Page similar to the reference application Login Page

## Issue I worked on
[RA-1469](https://issues.openmrs.org/browse/RA-1469)

## Checklist: I completed these to help reviewers
 - [x] My pull request only contains ONE single commit (the number above, next to the 'Commits' tab is 1).
No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [code style](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.
No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

 - [ ] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
No? -> execute above command

- [ ] All new and existing **tests passed**.
No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

 - [x] My pull request is **based on the latest changes** of the master branch.
No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: app/js/redux/actions/authActions.js [hunk 1]
```diff
@@ -89,3 +92,44 @@ export const createlogOutUrl = (locationResponse) => {
...
+};
+
+const loginUserSuccess = (dispatch, user) => {
+  dispatch({
+    type: actionTypes.LOGIN_USER_SUCCESS,
+    payload: user
+  });
+};
+
+const loginUserFail = (dispatch, error) => {
```
Comment (justmesam) [inline app/js/redux/actions/authActions.js]: there is an `asyncActionError` action which handles api call errors you should use that instead of creating another action to handle the errors

File: app/js/redux/reducers/authReducers.js [hunk 2]
```diff
@@ -19,39 +19,54 @@ export default (state=initialState.INITIAL_STATE, action) => {
...
     case actionTypes.ASYNC_ERROR:
       return {...state,
         loading : false,
         apiSyncError: action.error,
-      }
+      };
+    case actionTypes.LOGIN_USER:
+      return { ...state,
+         loading: true, 
+         error: ''
```
Comment (justmesam) [inline app/js/redux/reducers/authReducers.js]: if the `error` is not changed there is no need  of updating it to the new state

File: app/js/redux/actions/authActions.js [hunk 3]
```diff
@@ -89,3 +92,44 @@ export const createlogOutUrl = (locationResponse) => {
...
+      }
+    }).then(response => {
+      (response.data.authenticated) ? hashHistory.push('/home') : toastr.error('Invalid username or password');
+      loginUserSuccess(dispatch, response.data.user)
+    })
+      .catch(error => loginUserFail(dispatch, error));
+  }
+};
+
+const loginUserSuccess = (dispatch, user) => {
```
Comment (justmesam) [inline app/js/redux/actions/authActions.js]: the data that is returned from this api call is the same that is returned in the  fetch current user i think you should reuse `fetchSession` instead of creating a whole new method for the same thing

File: app/js/redux/actions/authActions.js [hunk 4]
```diff
@@ -89,3 +92,44 @@ export const createlogOutUrl = (locationResponse) => {
...
+      url: `${urlPrefix}/${apiBaseUrl}/${requestUrl}`, 
+      method: 'GET',
+      headers: {
+        'Authorization': authentication
+      }
+    }).then(response => {
+      (response.data.authenticated) ? hashHistory.push('/home') : toastr.error('Invalid username or password');
+      loginUserSuccess(dispatch, response.data.user)
+    })
+      .catch(error => loginUserFail(dispatch, error));
```
Comment (justmesam) [inline app/js/redux/actions/authActions.js]: this is not a bad implementation but i guess you could simplify it to something like
`.catch(error => dispatch(loginUserFail(error))` which is easier to read and you avoid passing to many  parameters to the method.

File: app/js/redux/constants/initialState.js [hunk 5]
```diff
@@ -4,4 +4,7 @@ export const INITIAL_STATE = {
   customLogoutURL: ' ',
   locationResponse: [],
   sessionResponse: [],
+  user: {},
+  loading: false,
+  error: ''
```
Comment (justmesam) [inline app/js/redux/constants/initialState.js]: add trailing comma

File: app/js/components/App.jsx [hunk 6]
```diff
@@ -11,41 +11,42 @@ import { connect } from "react-redux";
...
     return (
-      <div>
+      <div className="container main-container">
         <Header
           authentication={this.props.authentication}
           authActions={this.props}
-         />
-
+        />
+        <LoginForm />
```
Comment (justmesam) [inline app/js/components/App.jsx]: app should not render loginForm if  loginForm is supposed to be rendered before the app itself, since app carries the header which displays a user in session,
you could do two things ;
*   create auth component which renders the login form and redirects to app if login is successful
*   introduce a dashboard/home component which is a child component of app and it has the header and the app does conditional rendering if user is authenticated or not.

File: app/js/components/HomePage.jsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+import React from 'react';
+
+import Header from './presentational/common/Header/header.jsx';
+import * as AuthActions from '../redux/actions/authActions';
+
+
+class HomePage extends React.Component {
+
```
Comment (justmesam) [inline app/js/components/HomePage.jsx]: what is this component supposed to do????

File: app/js/components/LoginForm.jsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      username: '',
+      password: '',
+      sessionLocation: ''
+    };
+
+    this.handleChange = this.handleChange.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+  }
+
+  handleChange(event) {
```
Comment (justmesam) [inline app/js/components/LoginForm.jsx]: avoid using binding methods and conform to arrow functions to create uniformity in the whole app

File: app/js/components/LoginForm.jsx [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+import React from 'react';
+import { connect } from "react-redux";
+
+import { loginUser } from '../redux/actions/authActions';
+
+class LoginForm extends React.Component {
```
Comment (justmesam) [inline app/js/components/LoginForm.jsx]: it would be better if you made this a presentational component and create a container component for separation of concerns

File: app/js/redux/actions/authActions.js [hunk 10]
```diff
@@ -89,3 +92,44 @@ export const createlogOutUrl = (locationResponse) => {
...
+    dispatch({ type: actionTypes.LOGIN_USER });
+
+    const applicationDistribution = location.href.split('/')[2];
+    const urlPrefix = location.href.substr(0, location.href.indexOf('//'));
+    const url = location.href.split('/')[3];
+    const apiBaseUrl = `/${applicationDistribution}/${url}/ws/rest`;
+    const requestUrl = 'v1/session';
+    const authentication = 'Basic ' + new Buffer(`${username}:${password}`).toString('base64');
+
+    axios({
```
Comment (justmesam) [inline app/js/redux/actions/authActions.js]: there is an api helper which handles the configurations of the apihelper constant i think you should conform to use it.

File: app/js/routes.jsx [hunk 11]
```diff
@@ -6,16 +6,24 @@
  * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
  * graphic logo is a trademark of OpenMRS Inc.
  */
-import React from 'react'
-import {Route} from 'react-router'
+import React, { Component } from 'react';
+import { Router, Route, Link, IndexRoute, hashHistory, browserHistory } from 'react-router';
```
Comment (segunolalive) [inline app/js/routes.jsx]: Why is hashHIstory imported along with BrowserRouter?

File: app/js/redux/actions/authActions.js (hunk 1 above)
Comment (Charpell) [reply app/js/redux/actions/authActions.js]: Noted

File: app/js/redux/reducers/authReducers.js (hunk 2 above)
Comment (Charpell) [reply app/js/redux/reducers/authReducers.js]: Noted

File: app/js/redux/actions/authActions.js (hunk 3 above)
Comment (Charpell) [reply app/js/redux/actions/authActions.js]: Noted

File: app/js/redux/actions/authActions.js (hunk 4 above)
Comment (Charpell) [reply app/js/redux/actions/authActions.js]: Noted

File: app/js/redux/constants/initialState.js (hunk 5 above)
Comment (Charpell) [reply app/js/redux/constants/initialState.js]: Noted

File: app/js/components/App.jsx (hunk 6 above)
Comment (Charpell) [reply app/js/components/App.jsx]: The LoginForm would be rendered before the app and would not carry the header since the user is not authenticated.

File: app/js/components/HomePage.jsx (hunk 7 above)
Comment (Charpell) [reply app/js/components/HomePage.jsx]: Its a placeholder(Home Page) i used to redirect the user after a successful login.

File: app/js/components/LoginForm.jsx (hunk 8 above)
Comment (Charpell) [reply app/js/components/LoginForm.jsx]: Noted

File: app/js/components/LoginForm.jsx (hunk 9 above)
Comment (Charpell) [reply app/js/components/LoginForm.jsx]: Noted

File: app/js/redux/actions/authActions.js (hunk 10 above)
Comment (Charpell) [reply app/js/redux/actions/authActions.js]: Noted, the PR has been updated

File: app/js/routes.jsx (hunk 11 above)
Comment (Charpell) [reply app/js/routes.jsx]: PR has been updated

File: __tests__/actions/asynActions.test.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
+import configureMockStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+// import fetchMock from 'fetch-mock';
```
Comment (justmesam) [inline __tests__/actions/asynActions.test.js]: remove unused import

File: __tests__/actions/asynActions.test.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+describe('Async Actions', () => {
+  afterEach(() => {
+    nock.cleanAll();
+  })
+
+  it('should create LOGIN_USER and GET_CURRENT_SESSION when a user has logged in', (done) => {
+    const requestData = { username: 'admin', password: 'mypassword', sessionLocation: 'laboaratory' }
+    
+    nock('http://localhost:3002/openmrs-standalone/owa/openmrs-owa-referenceappui/index.html')
```
Comment (justmesam) [inline __tests__/actions/asynActions.test.js]: there is an `apihelper` its the same thing as this i would prefer you use it here

File: app/index.html [hunk 14]
```diff
@@ -3,20 +3,19 @@
...
   <!-- openmrs favicon -->
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
   <link rel="shortcut icon" type="image/ico" href="../../images/openmrs-favicon.ico"/>
   <link rel="icon" type="image/png" href="../../images/openmrs-favicon.png"/>
   <link rel="stylesheet" href="libs/bootstrap/dist/css/bootstrap.min.css" />
 <link rel="stylesheet" href="libs/toastr/toastr.min.css"/>
 <script src="libs/jquery/dist/jquery.min.js" ></script>
-<script src="libs/tether/dist/js/tether.mi" ></script>
+<script src="libs/tether/dist/js/tether.min.js" ></script>
```
Comment (justmesam) [inline app/index.html]: you should update the `.bowerrc` with every package you use...

File: app/index.html [hunk 15]
```diff
@@ -3,20 +3,19 @@
...
 <script src="libs/bootstrap/dist/js/bootstrap.min.js" ></script>
 <script src="libs/bootstrap-filestyle/src/bootstrap-filestyle.min.js" ></script>
 </head>
 <body>
-<div class="clear"></div>
-<div id="body-wrapper">
```
Comment (justmesam) [inline app/index.html]: why did you remove the class `body-wrapper` from the whole file?????

File: app/js/components/App.jsx [hunk 16]
```diff
@@ -11,41 +11,39 @@ import { connect } from "react-redux";
 
 import Header from './presentational/common/Header/header.jsx';
 import * as AuthActions from '../redux/actions/authActions';
+import Login from './Login';
+import omrsButton from './../../img/openmrs-with-title-small.png';
```
Comment (justmesam) [inline app/js/components/App.jsx]: use variable names which can be understood or descriptive  to their functions

File: app/js/components/presentational/Input.jsx [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+  return (
+    <input
+      className={className}
+      name={name}
+      type={type}
+      onChange={onChange}
+      value={value}
+    />
+  )
+}
```
Comment (justmesam) [inline app/js/components/presentational/Input.jsx]: add new line at the end

File: app/js/components/presentational/LoginForm.jsx [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+          <span className="forgot-password"> Forgot Password ?</span>
+        </div>
+        <button type="submit" className="login-button">Login</button>
+      </form>
+      </div>
+    </div>
+  )
+}
+
+export default LoginForm;
```
Comment (justmesam) [inline app/js/components/presentational/LoginForm.jsx]: add new line at the end

File: __tests__/actions/asynActions.test.js (hunk 13 above)
Comment (Charpell) [reply __tests__/actions/asynActions.test.js]: The nook test module expects a string

File: __tests__/components/Login.test.js [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+    loginUser: jest.fn()
+  }
+  const wrapper = mount(<Login  {...props} />);
+  return {
+    props,
+    wrapper
+  }
+}
+
+global.event = {
```
Comment (segunolalive) [inline __tests__/components/Login.test.js]: attaching an event to the global scope can cause unexpected side effects that can be difficult to debug as the project grows. Simply use a local variable

File: __tests__/actions/asynActions.test.js (hunk 13 above)
Comment (Charpell) [reply __tests__/actions/asynActions.test.js]: The apiHelper is a function and won't work in this instance

File: __tests__/components/Login.test.js (hunk 19 above)
Comment (Charpell) [reply __tests__/components/Login.test.js]: I have updated the PR

File: app/index.html (hunk 15 above)
Comment (Charpell) [reply app/index.html]: The PR has been updated

File: app/js/components/presentational/Input.jsx [hunk 20]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+
+export default Input = ({ className, name, type, onChange, value }) => {
```
Comment (justmesam) [inline app/js/components/presentational/Input.jsx]: I cant see anywhere you are using this component

Comment (Charpell) [reply app/js/components/presentational/Input.jsx]: PR has been updated
